src/lib/utils.ts

```diff
@@ -83,29 +85,165 @@ export const promiseErrorMultiplier = <T>(
   });
 };
 
-export const getL1GatewayAddress = async (l2GatewayAddress: string) => {
-  const l2Gateway = isNova
-    ? l2ToL1GatewayAddressesNova[l2GatewayAddress.toLowerCase()]
-    : l2ToL1GatewayAddresses[l2GatewayAddress.toLowerCase()];
-
-  if (l2Gateway) return l2Gateway;
-
-  return undefined;
-
-  // TODO: discuss:
-  // try {
-  //   const tokenGateway = TokenGateway__factory.connect(
-  //     l2GatewayAddress,
-  //     l2Provider
-  //   );
-  //   const l1Gateway = await promiseErrorMultiplier(
-  //     tokenGateway.counterpartGateway(),
-  //     (error) => tokenGateway.counterpartGateway()
-  //   );
-  //   return l1Gateway;
-  // } catch (e) {
-  //   return undefined;
-  // }
+export const generateGatewayMap = async (
+  l2Multicaller: MultiCaller,
+  l2Network: L2Network,
+  l1Provider: Provider
+) => {
+  const l1GatewayResults: Map<string, string> = new Map();
+  const l1Token: any[] = [];
+  const gatewayMap: Map<string, string> = new Map();
+
+  //default gateway can be set during initialize call, it does not emit GatewaySet, so we should
+  //manually set it
+  {
+    const l1GatewayRouter = L1GatewayRouter__factory.connect(
+      l2Network.tokenBridge.l1GatewayRouter,
+      l1Provider
+    );
+    const defaultGateway = await l1GatewayRouter.defaultGateway();
+    const defaultGatewayContract = await new ethers.Contract(
+      defaultGateway,
+      TokenGateway__factory.abi
+    ).connect(l1Provider);
+    const defaultCounterPartGateway =
+      await defaultGatewayContract.counterpartGateway();
+    gatewayMap.set(
+      defaultCounterPartGateway.toLowerCase(),
+      defaultGateway.toLowerCase()
+    );
+  }
+
+  const gatewaySetsList = await getGatewaysets()
+  
+  for(let i = 0; i < gatewaySetsList.length; i++) {
+    const tokenAddress = gatewaySetsList[i].l1Token
+    let l1GatewayAddress = gatewaySetsList[i].gateway
+
+    //if gateway set to zero, which means it sets back to standard erc20 gateway
+    if (l1GatewayAddress === ethers.constants.AddressZero) {
+      l1GatewayAddress = l2Network.tokenBridge.l1ERC20Gateway;
+    }
+    l1GatewayResults.set(tokenAddress, l1GatewayAddress);
+    l1Token.push(tokenAddress);
+  }
+  
+  const l2GatewayMaps = await getL2GatewayAddressesFromL1Token(
+    l1Token,
+    l2Multicaller,
+    l2Network
+  );
+
+  //set gateway map
+  for (let i = 0; i < l1Token.length; i++) {
+    if (l2GatewayMaps[i] === ethers.constants.AddressZero) continue;
+    gatewayMap.set(
+      l2GatewayMaps[i].toLowerCase(),
+      l1GatewayResults.get(l1Token[i])!.toLowerCase()
+    );
+  }
+
+  //edge case: gateway registered on l1 while not on l2
+  if (!(await checkMapResultByL2Gateway(gatewayMap, l2Multicaller))) {
+    exit(1);
```

Comment:
print some error before exiting would be nice

Reply:
It print at https://github.com/OffchainLabs/arb-token-lists/blob/dynamic-gateway/src/lib/utils.ts#L173